### DISCUSSION TITLE
Anemometer using a Reed-contact

### DISCUSSION THREAD

**Discussion by azakow** [arendst/Tasmota#11500] [Q&A] [answered]
Hello all,

I like to build a weather station using Tasmota. The tasmota-sensors.bin is a good start for many weather attribute (BH1750, BME280, AS3935). I understand the tasmota-sensors.bin is lacking support of a anemometer using a Reed-contact.

Looking through github I found this [xsns_68_windmeter.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_68_windmeter.ino) From the beginning of the code (diameter, pulses per rotation …) I understand that it could be adjusted to any given anemometer size using a Reed-contact.

I understand that I could build my own bin-file with Tasmota to support what I plan to do. My programming/compiling skills are very limited. Compiling bin-files is something I do have great respect since I assume I could kill my components or do not have the skills to debug.

Looking here [Compile-your-build](https://tasmota.github.io/docs/Compile-your-build/) it seems possible to integrate my own code into Tasmota, but “own” code seems way beyond what I plan to do, since the "required" code is already there [xsns_68_windmeter.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_68_windmeter.ino) and needs to be “integrated” only.

Now before I start a journey on my own and maybe get lost: 
Please let me know, which steps do I have to perform to create my own Tasmota-bin-file supporting the following sensors: BH1750, BME280, AS3935 **and** integrate xsns_68_windmeter.ino.
Thank you.

**Comment by Jason2866** [comment, marked as answer]
You have to compile yourself a version with the drivers / sensors you need.
It sounds more sophisticated than it is. Read https://tasmota.github.io/docs/Compile-your-build/
how to do

**Reply by azakow** [reply]
So all I need to show is courage and follow the guide!
1. setup a local environment
2. edit the file "my_user_config.h" to `#define USE_WINDMETER            // Add support for analog anemometer (+2k2 code)`,
4. compile,
5. upload to the device,
6. done.

I read that available memory can be an issue. Will adding the analog anemometer cause issues?

**Reply by Jason2866** [reply]
Do not edit `my_user_config.h` rename `user_config_override_sample.h` to ` user_config_override.h`
and do add the `#define xxx...` here.
Compile, upload and done.
Dont worry about the flash size and ram. Enough left

**Reply by azakow** [reply]
I understand to cover my requirements I would add the following to `user_config_override.h`
```
// Add support for analog anemometer (+2k2 code)
#ifndef USE_WINDMETER
#define USE_WINDMETER       
#endif
```

I like to use more sensors with the target hardware, like BME280, BH1750 and AS3935.
Do these have to be added in the same manner to make them available to the resulting bin-file?
```
// Add support for analog anemometer (+2k2 code)
#ifndef USE_WINDMETER
#define USE_WINDMETER
#endif

// [I2cDriver10] Enable BMP085/BMP180/BMP280/BME280 sensors (I2C addresses 0x76 and 0x77) (+4k4 code)
#ifndef USE_BMP
#define USE_BMP
#endif

// [I2cDriver11] Enable BH1750 sensor (I2C address 0x23 or 0x5C) (+0k5 code)
#ifndef USE_BH1750
#define USE_BH1750
#endif

// [I2cDriver48] Enable AS3935 Franklin Lightning Sensor (I2C address 0x03) (+5k4 code)
#ifndef USE_AS3935
#define USE_AS3935
#endif
```

**Reply by barbudor** [reply]
Yes, you can add all the stuff you need.
Just build "tasmota" and not any of the other pre-defined binary.
Each predefined binary may include optimization tricks or specifics that may fail if you add features.

**Reply by sfromis** [reply]
Most of the time, trying add extra features to the preconfigured extra builds will just be ignored.

**Reply by azakow** [reply]
It worked surprisingly well to compile my very 1st own flashware.  Mind I am a total noob! 👯‍♂️
Checking the selectable options for a GPIO pin, I can now choose "WindMeter Spd".  👯‍♂️ 

Next stop: Setting the parameters for my anemometer via the console for **Sensor68**.

This is great stuff!
Thank you all!

![2021-03-29 21_47_28-Tasmota - Main Menu](https://user-images.githubusercontent.com/79204427/112891792-d8d36380-90d8-11eb-9c47-1b8160016318.png)
